### PR TITLE
Fix cart rule shop association

### LINKED ISSUE #28148: Shop associations for cart_rules are saved without taking into account the employee permissions
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

In a multishop environment, when selecting all available shop in Shop selection, nothing is saved to `ps_cart_rule_shop`, or if there are entries, they are deleted. This is correct if the employee has access to all shop. It is wrong if not.

**Case 1:**
An employee has access to only one shop. If he creates a cart rule, it is valid in all shops. There is no way to avoid this behavior.
This can have disastrous consequences: If the employee does not understand what he does, he can offer free orders in any random shop. If shops are independent, this is a nightmare.

**Case 2:**
A cart rule is limited to one shop (created by an employee with permissions for more than one shop).
An other employee, that has only permissions in that shop, views the cart rule and saves it without changes.
Now the cart rule is corrupted, and applies to all shops.

**Case 3:**
A cart rule is limited to one shop. An employee with access to two shops adds the second shop and saves it. Now the cart rule is valid in all shops, and the checkbox `shop restriction` has been removed (what represents an error).

### Expected behavior

I expect, that the shop associations of a cart rule reflect the permissions of the creator of the rule. There must be no possible permission escalation.

**Case 1:**
If an employee is restricted to a set of shops, all cart rules he creates must be restricted to shops inside that set.

**Case 2:**
If a cart rule is limited to a set of shops, these shop associations must not change, unless there is a explicit action by the user.

**Case 3:**
If `Shop restriction` is checked, it must not be removed. A cart rule associated to two shops must remain associated to the two shops, unless there is a explicit action by the user.

### Steps to reproduce

Setup:
- Multishop environment with at least three shops
- Edit or create a employee profile with permissions for cart rules, and use this profile for creating two employees as follows:
- An employee A that has permissions only for shop A
- An employee B that has permissions for shop A and B

**Case 1:**
1. Log in as employee A
2. Create a cart rule
3. Note that there is no option to select `Shop restriction` in the conditions tab 
image
4. Log in as super admin
5. Find the new cart rule, click on `Edit`
6. In the conditions tab > note that the cart rule has no `Shop restriction` and is valid in all shops.

https://user-images.githubusercontent.com/16067358/162452982-80e2717a-bb57-4b4b-9878-550c0c705eae.mp4

**Case 2:**
1. Log in as super admin
2. Create a cart rule restricted to shop A
3. Log in as user A
4. View the cart rule, do not make any modification, and click on `Save`
5. Log in as super admin
6. Note that the shop restriction has been removed from the cart rule

**Case 3:**
1. Log in as user B
2. Create a cart rule restricted to shop A
3. Edit the cart rule
4. Add shop B to the shop restrictions, click `Save and stay`
5. Note that `Shop restrictions` has been unchecked
6. You can note the same thing as super admin

### PrestaShop version(s) where the bug happened

1.6.1.1, 1.7.7.1, 1.7.8.5

### PHP version(s) where the bug happened

7.3

### If your bug is related to a module, specify its name and its version

_No response_

### PR DESCRIPTION
**Pull request by lmeyer1**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | The shop association of a cart rule is filled automatically, if an employee is restricted to certain shops and does not select himself shop associated to the cart rule. Never uncheck `shop association` if a shop is already selected. If we check a simple restriction, but do not select an item, an error is raised.
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #28148.
| Related PRs       | Related to, but independent from #28145
| How to test?      | All three cases explained in #28148 are no longer possible to reproduce
| Possible impacts? | Permissions are locked down. An employee restricted to a set of shops can no longer create cart rules valid in other shops

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/28153)

### CODE REVIEW

**Comment by kpodemski** [comment]
Hello @lmeyer1 - thanks for your PR.

This is not regression. I think you should target the `develop` branch.

File: admin-dev/themes/default/template/controllers/cart_rules/conditions.tpl [hunk 1]
```diff
@@ -294,10 +294,10 @@
 				</a>
 			</div>
 
-		{if ($shops.unselected|@count) + ($shops.selected|@count) > 1}
+		{if ($all_shops|@count) > 1}
 			<p class="checkbox">
 				<label>
-					<input type="checkbox" id="shop_restriction" name="shop_restriction" value="1" {if $shops.unselected|@count}checked="checked"{/if} />
+					<input type="checkbox" id="shop_restriction" name="shop_restriction" value="1" {if $shops.selected|@count || $shops.selected|@count + $shops.unselected|@count < $all_shops|@count}checked="checked"{/if} />
```
Comment (Progi1984) [inline admin-dev/themes/default/template/controllers/cart_rules/conditions.tpl]:
```suggestion
					<input type="checkbox" id="shop_restriction" name="shop_restriction" value="1" {if $shops.selected|@count || ($shops.selected|@count + $shops.unselected|@count < $all_shops|@count)}checked="checked"{/if} />
```

File: controllers/admin/AdminCartRulesController.php [hunk 2]
```diff
@@ -297,6 +319,21 @@ public function processAdd()
...
     {
+        // Add shop restrictions if employee has not access to all shops
+        $context = Context::getContext();
+        $all_shops = Shop::getCompleteListOfShopsID();
+        if ($context->employee->isSuperAdmin()) {
+            $employee_shops = $all_shops;
+        } else {
+            $employee_shops = $context->employee->getAssociatedShops();
+        }
+        if (count($all_shops) > count($employee_shops) && Tools::getValue('shop_restriction') == '0') {
```
Comment (Progi1984) [inline controllers/admin/AdminCartRulesController.php:330]:
```suggestion
        if (!empty($employee_shops) && count($all_shops) > count($employee_shops) && Tools::getValue('shop_restriction') == '0') {
```

**Review by Progi1984 — COMMENTED** [review]
Small feedbacks

File: controllers/admin/AdminCartRulesController.php (hunk 2 above)
Comment (lmeyer1) [reply controllers/admin/AdminCartRulesController.php:330]: `$employee_shops` is guaranteed to be an array. And it cannot ever be empty.

File: controllers/admin/AdminCartRulesController.php [hunk 3]
```diff
@@ -200,13 +200,35 @@ public function postProcess()
...
+                            break;
+                        case 'group':
+                            $restriction_name = $this->trans('Customer group selection', [], 'Admin.Catalog.Feature');
+                            break;
+                        case 'shop':
+                        default:
+                            $restriction_name = $this->trans('Shop selection', [], 'Admin.Catalog.Feature');
+                            break;
+                    }
+                    $this->errors[] = $this->trans("'%s' is checked, but no item selected", [$restriction_name], 'Admin.Catalog.Notification');
```
Comment (Julievrz) [inline controllers/admin/AdminCartRulesController.php]: Hi @lmeyer1, 
I'm not sure to understand what is behind this variable. And what do you mean by "item"? Do you mean the restriction checkboxes?

Comment (lmeyer1) [reply controllers/admin/AdminCartRulesController.php]: @Julievrz 
Logically, if someone selects the restriction checkbox, e.g. for Countries, but does not select any country, there is a problem.
Until now, in this case, the restriction checkbox was disregarded, and the cart rule applied to all countries.
But disregard the checkbox, even though it is checked was part of the bugs, I'm fixing here.
"'%s' is checked, but no item selected" becomes "'Country selection' is checked, but no item selected", where item is a country in the list.

Comment (Julievrz) [inline controllers/admin/AdminCartRulesController.php]:
```suggestion
                    $this->errors[] = $this->trans(' The "%s" restriction is checked but no item is selected.", [$restriction_name], 'Admin.Catalog.Notification');
```

Comment (Julievrz) [reply controllers/admin/AdminCartRulesController.php]: Thanks, it's clearer now. :-)

File: admin-dev/themes/default/template/controllers/cart_rules/conditions.tpl [hunk 4]
```diff
@@ -294,10 +294,10 @@
 				</a>
 			</div>
 
-		{if ($shops.unselected|@count) + ($shops.selected|@count) > 1}
+		{if ($all_shops|@count) > 1}
 			<p class="checkbox">
 				<label>
-					<input type="checkbox" id="shop_restriction" name="shop_restriction" value="1" {if $shops.unselected|@count}checked="checked"{/if} />
+					<input type="checkbox" id="shop_restriction" name="shop_restriction" value="1" {if $shops.selected|@count || ($shops.selected|@count + $shops.unselected|@count < $all_shops|@count)}checked="checked"{/if} />
```
Comment (Progi1984) [inline admin-dev/themes/default/template/controllers/cart_rules/conditions.tpl:300]:
```suggestion
					<input type="checkbox" id="shop_restriction" name="shop_restriction" value="1" {if $shops.selected|@count || (($shops.selected|@count + $shops.unselected|@count) < $all_shops|@count)}checked="checked"{/if} />
```

**Review by Progi1984 — COMMENTED** [review]
A small improvement for better readibility

File: admin-dev/themes/default/template/controllers/cart_rules/conditions.tpl (hunk 4 above)
Comment (lmeyer1) [reply admin-dev/themes/default/template/controllers/cart_rules/conditions.tpl:300]: You already made a suggestion on this line https://github.com/PrestaShop/PrestaShop/pull/28153#discussion_r848056768, where I fully agreed. 
But for this suggested change, I think everybody knows how to read `a + b < c`, therefore parenthesis `(a + b) < c` are superfluous here

**Review by atomiix — APPROVED** [review]

**Review by Progi1984 — APPROVED** [review]

**Review by Julievrz — APPROVED** [review]
Wording ✅

**Review by AureRita — APPROVED** [review]
Hi @lmeyer1 

Thank you for your PR, I tested it and all case explain works, as you can see :

https://user-images.githubusercontent.com/100699445/166462839-3afbcdbb-bbde-44bf-bb5c-49ec05508ddb.mp4

**Comment by Progi1984** [comment]
Thanks @lmeyer1 & @AureRita

**Pull request merged**